File: src/__tests__/stallGuard.test.ts
/**
 * __tests__/stallGuard.test.ts
 * Unit tests for safety/stallGuard.ts – stall detection and fallback behaviour.
 */
import { StallGuard } from '../safety/stallGuard';

describe('StallGuard – basic recording', () => {
  it('is not stalled before any attempts', () => {
    const guard = new StallGuard();
    expect(guard.isStalled('craft_tools')).toBe(false);
  });

  it('detects stall after maxAttempts failures', () => {
    const guard = new StallGuard({ maxAttempts: 3, timeoutMs: 60_000 });
    guard.recordAttempt('craft_tools');
    guard.recordAttempt('craft_tools');
    expect(guard.isStalled('craft_tools')).toBe(false);
    guard.recordAttempt('craft_tools');
    expect(guard.isStalled('craft_tools')).toBe(true);
  });

  it('clears stall on success', () => {
    const guard = new StallGuard({ maxAttempts: 2, timeoutMs: 60_000 });
    guard.recordAttempt('gather_food');
    guard.recordAttempt('gather_food');
    expect(guard.isStalled('gather_food')).toBe(true);
    guard.recordSuccess('gather_food');
    expect(guard.isStalled('gather_food')).toBe(false);
  });

  it('detects stall after timeout', () => {
    // Use a very short timeout
    const guard = new StallGuard({ maxAttempts: 100, timeoutMs: 0 });
    guard.recordAttempt('build_shelter');
    expect(guard.isStalled('build_shelter')).toBe(true);
  });
});

describe('StallGuard – fallback actions', () => {
  const guard = new StallGuard();

  it('returns prerequisite for craft_tools → gather_wood', () => {
    expect(guard.getFallbackAction('craft_tools')).toBe('gather_wood');
  });

  it('returns prerequisite for build_shelter → gather_wood', () => {
    expect(guard.getFallbackAction('build_shelter')).toBe('gather_wood');
  });

  it('returns explore for unknown subgoal', () => {
    expect(guard.getFallbackAction('unknown_exotic_action')).toBe('explore');
  });

  it('handles compound role subgoal ids (extracts last segment)', () => {
    // 'role_farmer_food' → last segment 'food' → not in map → 'explore'
    expect(guard.getFallbackAction('role_farmer_food')).toBe('explore');
  });
});

describe('StallGuard – reset', () => {
  it('clears all records on reset', () => {
    const guard = new StallGuard({ maxAttempts: 1, timeoutMs: 60_000 });
    guard.recordAttempt('gather_wood');
    guard.recordAttempt('eat_food');
    expect(guard.getRecords()).toHaveLength(2);
    guard.reset();
    expect(guard.getRecords()).toHaveLength(0);
  });
});

File: src/test/java/io/github/officiallymikey7/mini/StallGuardTest.java
package io.github.officiallymikey7.mini;

import io.github.officiallymikey7.mini.safety.StallGuard;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for {@link StallGuard}.
 *
 * <p>Mirrors the TypeScript {@code stallGuard.test.ts} suite.
 */
class StallGuardTest {

    private StallGuard guard;

    @BeforeEach
    void setUp() {
        // maxAttempts=3, timeoutMs=60_000
        guard = new StallGuard(3, 60_000);
    }

    // ── Not stalled initially ─────────────────────────────────────────────────

    @Test
    void notStalledWithNoAttempts() {
        assertFalse(guard.isStalled("some_goal"));
    }

    @Test
    void notStalledAfterOneAttempt() {
        guard.recordAttempt("goal_a");
        assertFalse(guard.isStalled("goal_a"));
    }

    @Test
    void notStalledAfterTwoAttempts() {
        guard.recordAttempt("goal_a");
        guard.recordAttempt("goal_a");
        assertFalse(guard.isStalled("goal_a"));
    }

    // ── Stalled after maxAttempts ─────────────────────────────────────────────

    @Test
    void stalledAfterThreeAttempts() {
        guard.recordAttempt("goal_a");
        guard.recordAttempt("goal_a");
        guard.recordAttempt("goal_a");
        assertTrue(guard.isStalled("goal_a"));
    }

    // ── Success clears stall ──────────────────────────────────────────────────

    @Test
    void successClearsStall() {
        guard.recordAttempt("goal_a");
        guard.recordAttempt("goal_a");
        guard.recordAttempt("goal_a");
        assertTrue(guard.isStalled("goal_a"));
        guard.recordSuccess("goal_a");
        assertFalse(guard.isStalled("goal_a"));
    }

    // ── Reset clears all records ──────────────────────────────────────────────

    @Test
    void resetClearsAllRecords() {
        guard.recordAttempt("goal_a");
        guard.recordAttempt("goal_b");
        guard.reset();
        assertTrue(guard.getRecords().isEmpty());
        assertFalse(guard.isStalled("goal_a"));
    }

    // ── Fallback actions ──────────────────────────────────────────────────────

    @Test
    void fallbackForCraftTools() {
        assertEquals("gather_wood", guard.getFallbackAction("craft_tools"));
    }

    @Test
    void fallbackForBuildShelter() {
        assertEquals("gather_wood", guard.getFallbackAction("build_shelter"));
    }

    @Test
    void fallbackForEatFood() {
        assertEquals("forage_food", guard.getFallbackAction("eat_food"));
    }

    @Test
    void fallbackForUnknownGoal() {
        assertEquals("explore", guard.getFallbackAction("some_unknown_goal"));
    }

    @Test
    void fallbackForCompoundId_roleFarmerFood() {
        // "role_farmer_food" – base segment is "food", no mapping → explore
        assertEquals("explore", guard.getFallbackAction("role_farmer_food"));
    }

    // ── Multiple goals independently tracked ─────────────────────────────────

    @Test
    void independentGoals() {
        guard.recordAttempt("goal_a");
        guard.recordAttempt("goal_a");
        guard.recordAttempt("goal_a"); // stalled

        guard.recordAttempt("goal_b");

        assertTrue(guard.isStalled("goal_a"));
        assertFalse(guard.isStalled("goal_b"));
    }

    // ── Timeout-based stall ───────────────────────────────────────────────────

    @Test
    void timeoutStall() throws InterruptedException {
        StallGuard fastGuard = new StallGuard(100, 50); // 50ms timeout, 100 max attempts
        fastGuard.recordAttempt("goal_slow");
        assertFalse(fastGuard.isStalled("goal_slow"));
        Thread.sleep(150); // exceed 50ms timeout with ample margin for scheduler jitter
        assertTrue(fastGuard.isStalled("goal_slow"));
    }
}
